### PR TITLE
Lite's collectives feature

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes #\<issue_number>

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/lightning_lite/plugins/collectives/collective.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
+import datetime
+from abc import ABC, abstractmethod
+from typing import Any, List, Optional
+
+import torch
+from torch.distributed import ProcessGroup, ReduceOp
+
+
+class Collective(ABC):
+    def __init__(self, delayed_init: bool, *init_args, **init_kwargs):
```
Comment (carmocca) [inline src/lightning_lite/plugins/collectives/collective.py:10]:
```suggestion
    def __init__(self, instantiate_pg: bool, *pg_args, **pg_kwargs):
```

File: src/lightning_lite/plugins/collectives/collective.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+        if delayed_init:
+            self._group = None
+            self._init_args = init_args
+            self._init_kwargs = init_kwargs
+        else:
+            self._group = self._init_group_impl(*init_args, **init_kwargs)
```
Comment (carmocca) [inline src/lightning_lite/plugins/collectives/collective.py:16]: we should define these unconditionally.

And we can rely on the getter for instantaition

```suggestion
        self._group = None
        self._init_args = init_args
        self._init_kwargs = init_kwargs
        if intantiate_pg:
            _ = self.group 
```

File: src/lightning_lite/plugins/collectives/collective.py (hunk 1 above)
Comment (carmocca) [inline src/lightning_lite/plugins/collectives/collective.py:10]: I would consider only supporting kwargs to make it clear what everything is

File: src/lightning_lite/plugins/collectives/collective.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+    @property
+    def group(self) -> ProcessGroup:
+        if self._group is None:
+            self._group = self._init_group_impl(*self._init_args, **self._init_kwargs)
+        return self._group
+
+    @abstractmethod
+    def _init_group_impl(
```
Comment (carmocca) [inline src/lightning_lite/plugins/collectives/collective.py:25]: Why protected?

```suggestion
    def init_group(
```

File: src/lightning_lite/plugins/collectives/collective.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+
+class Collective(ABC):
+    def __init__(self, instantiate_group: bool = False, **group_kwargs: Any) -> None:
+        self._group_kwargs = group_kwargs
+        self._group: Optional[CollectibleGroup] = None
+        if instantiate_group:
+            self.create_group()
+
+    def create_group(self, **kwargs: Any) -> Self:  # type: ignore[valid-type]
```
Comment (carmocca) [inline src/lightning_lite/plugins/collectives/collective.py:24]: TODO: we need to reorder the properties, methods, abstractmethods for all classes eventually

File: src/lightning_lite/plugins/collectives/deepspeed_collective.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        **kwargs: Any,
+    ) -> ProcessGroup:
+        return dist.init_process_group(**kwargs)
+
+    def broadcast(
+        self,
+        tensor: torch.Tensor,
+        src: int,
+    ) -> torch.Tensor:
+        torch.distirbuted.broadcast(tensor, src, group=self.group)
```
Comment (carmocca) [inline src/lightning_lite/plugins/collectives/deepspeed_collective.py:31]: Wrong use

File: src/lightning_lite/plugins/collectives/__init__.py [hunk 6]
```diff
@@ -0,0 +1,11 @@
+from lightning_lite.plugins.collectives.collective import Collective
+from lightning_lite.plugins.collectives.deepspeed_collective import DeepSpeedCollective
+from lightning_lite.plugins.collectives.single_device_collective import SingleDeviceCollective
+from lightning_lite.plugins.collectives.torch_collective import TorchCollective
+
+__all__ = [
+    "Collective",
+    "DeepSpeedCollective",
+    "TorchCollective",
+    "SingleDeviceCollective",
```
Comment (carmocca) [inline src/lightning_lite/plugins/collectives/__init__.py:10]: XLA missing

File: src/lightning_lite/plugins/collectives/deepspeed_collective.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
+import datetime
+from typing import Any, List, Optional
+
+import torch
+
+from lightning_lite.plugins.collectives.collective import Collective
+from lightning_lite.strategies.deepspeed import _DEEPSPEED_AVAILABLE
+from lightning_lite.utilities.types import CollectibleGroup
+
+if _DEEPSPEED_AVAILABLE:
```
Comment (carmocca) [inline src/lightning_lite/plugins/collectives/deepspeed_collective.py:10]: In the spirit of #12786, we should make these local imports
